### PR TITLE
.Net: Added implementation of Weaviate connector for new memory design

### PR DESCRIPTION
**Pull request by dmytrostruk**
### Motivation and Context

Related: https://github.com/microsoft/semantic-kernel/issues/6527

In this PR:
- Implemented `IVectorStore`
- Implemented `IVectorStoreRecordCollection<TKey, TRecord>`
- Weaviate default record mapper
- `Options` classes
- Extension methods for DI
- Integration tests  
- Unit tests

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStore.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+/// <summary>
+/// Class for accessing the list of collections in a Weaviate vector store.
+/// </summary>
+/// <remarks>
+/// This class can be used with collections of any schema type, but requires you to provide schema information when getting a collection.
+/// </remarks>
+public sealed class WeaviateVectorStore : IVectorStore
+{
+    /// <summary><see cref="HttpClient"/> that can be used to manage the collections in Weaviate.</summary>
```
Comment (SergeyMenshykh) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStore.cs:22]: Super nit :) The http client is not used for collection management, right? It's for accessing the store API?

`/// <summary>An instance of <see cref="HttpClient"/> used to interact with Weaviate's API.</summary>`

**Review by SergeyMenshykh — APPROVED** [review]

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/Converters/WeaviateDateTimeOffsetConverter.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    public override DateTimeOffset Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    {
+        var dateString = reader.GetString();
+
+        if (string.IsNullOrWhiteSpace(dateString))
+        {
+            return default;
+        }
+
+        return DateTimeOffset.Parse(dateString);
```
[the remark this answers is not in the document]
Comment (dmytrostruk) [reply dotnet/src/Connectors/Connectors.Memory.Weaviate/Converters/WeaviateDateTimeOffsetConverter.cs]: Good point, added `CultureInfo.InvariantCulture`.

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateKernelBuilderExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Net.Http;
+using Microsoft.SemanticKernel.Data;
+
+namespace Microsoft.SemanticKernel.Connectors.Weaviate;
```
[the remark this answers is not in the document]
Comment (dmytrostruk) [reply dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateKernelBuilderExtensions.cs]: Ahh, good catch, thank you! Updated.

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStore.cs (hunk 1 above)
Comment (dmytrostruk) [reply dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStore.cs]: It's for accessing the store API for collection management :) But I've updated the documentation to better explain the purpose of `HttpClient` in this class.

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStore.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+        return recordCollection!;
+    }
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<string> ListCollectionNamesAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        using var request = new WeaviateGetCollectionsRequest().Build();
+
+        var response = await this._httpClient.SendWithSuccessCheckAsync(request, cancellationToken).ConfigureAwait(false);
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStore.cs:75]: Should there be support for passing an API key with this request?

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStore.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        var responseContent = await response.Content.ReadAsStringWithExceptionMappingAsync().ConfigureAwait(false);
+        var collectionResponse = JsonSerializer.Deserialize<WeaviateGetCollectionsResponse>(responseContent);
+
+        if (collectionResponse?.Collections is not null)
+        {
+            foreach (var collection in collectionResponse.Collections)
+            {
+                if (!string.IsNullOrWhiteSpace(collection.CollectionName))
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStore.cs:83]: Can this happen and if yes is this expected or an error state?

The behaviour here is to ignore the invalid collection name, I dont think we should make this decision for the host application, consider either passing the collection name to the next level as-is or throwing.

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreCollectionCreateMapping.cs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.SemanticKernel.Data;
+
+namespace Microsoft.SemanticKernel.Connectors.Weaviate;
+
+internal static class WeaviateVectorStoreCollectionCreateMapping
+{
+    public static WeaviateCollectionSchema MapToSchema(
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreCollectionCreateMapping.cs:13]: nit: Add some API documentation to explain these utility methods

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreCollectionCreateMapping.cs [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        {
+            return Hnsw;
+        }
+
+        return indexKind switch
+        {
+            IndexKind.Hnsw => Hnsw,
+            IndexKind.Flat => Flat,
+            IndexKind.Dynamic => Dynamic,
+            _ => throw new InvalidOperationException($"Index kind '{indexKind}' on {nameof(VectorStoreRecordVectorProperty)} '{vectorPropertyName}' is not supported by the Weaviate VectorStore.")
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreCollectionCreateMapping.cs:69]: List the index kinds that are supported in the exception message

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreCollectionCreateMapping.cs [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        }
+
+        return distanceFunction switch
+        {
+            DistanceFunction.CosineDistance => Cosine,
+            DistanceFunction.DotProductSimilarity => Dot,
+            DistanceFunction.EuclideanSquaredDistance => EuclideanSquared,
+            DistanceFunction.Hamming => Hamming,
+            DistanceFunction.ManhattanDistance => Manhattan,
+            _ => throw new InvalidOperationException($"Distance function '{distanceFunction}' on {nameof(VectorStoreRecordVectorProperty)} '{vectorPropertyName}' is not supported by the Weaviate VectorStore.")
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreCollectionCreateMapping.cs:94]: List the distance functions that are supported in the exception message

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreRecordCollection.cs [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+        string collectionName,
+        WeaviateVectorStoreRecordCollectionOptions<TRecord>? options = null)
+    {
+        // Verify.
+        Verify.NotNull(httpClient);
+        Verify.NotNullOrWhiteSpace(collectionName);
+
+        // Assign.
+        this._httpClient = httpClient;
+        this.CollectionName = collectionName;
```
Comment (atiq-bs23) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreRecordCollection.cs:129]: @dmytrostruk 
To align with Weaviate's naming conventions, it might be helpful to sanitize the collection name. Implemented in v1
[Documentation link](https://weaviate.io/developers/weaviate/manage-data/collections#class)

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreCollectionCreateMapping.cs [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.SemanticKernel.Data;
+
+namespace Microsoft.SemanticKernel.Connectors.Weaviate;
+
+internal static class WeaviateVectorStoreCollectionCreateMapping
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreCollectionCreateMapping.cs:11]: We should aim for high code coverage for this class, can you check we have that

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreRecordCollection.cs [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+    public string CollectionName { get; }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="WeaviateVectorStoreRecordCollection{TRecord}"/> class.
+    /// </summary>
+    /// <param name="httpClient"><see cref="HttpClient"/> that is used to interact with Weaviate API.</param>
+    /// <param name="collectionName">The name of the collection that this <see cref="WeaviateVectorStoreRecordCollection{TRecord}"/> will access.</param>
+    /// <param name="options">Optional configuration options for this class.</param>
+    public WeaviateVectorStoreRecordCollection(
+        HttpClient httpClient,
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreRecordCollection.cs:119]: Curious about authentication here also?

If the `HttpClient` needs to be configured with authentication options then we should make this very clear in the API documentation. An option to pass in the API key might be nice also as we do this elsewhere in our API.

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreRecordCollection.cs (hunk 9 above)
Comment (dmytrostruk) [reply dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreRecordCollection.cs:129]: @atiq-bs23 Thanks for your comment. I'm not sure we need to perform name sanitization on Semantic Kernel side. In my opinion, users could handle naming logic and follow Weaviate rules. A couple of reasons for that:
1. If user will provide one name, but collection will be silently created with another sanitized name, there could be some questions why the name was changed. Also, our sanitization rules may not work for all users, so they still be forced to implement their own sanitization logic if needed, so having this logic on our side won't fix the problem entirely. 
2. Naming conventions and rules in Weaviate can change over time, which means that we will have to update the logic every time there are some changes on Weaviate side. In this case, it won't allow users to proceed with this connector until we push new changes. On the other hand, if users will be responsible to provide final valid collection name, they will also have the ability to change it anytime on their side if there are some changes in Weaviate. So, it's more flexible.

In short, it's better to use the name as it's provided in constructor. If it's incorrect, user will receive an exception and will be able to update the name according to Weaviate naming rules.

Comment (atiq-bs23) [reply dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreRecordCollection.cs:129]: Thanks for clarify.

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStore.cs (hunk 4 above)
Comment (dmytrostruk) [reply dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStore.cs:81]: Added a possibility to specify API key in `options` and added XML documentation for `HttpClient` that it can also be included via `DefaultRequestHeaders` property. Also added tests to verify it is used correctly.

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStore.cs (hunk 5 above)
Comment (dmytrostruk) [reply dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStore.cs]: I don't expect this to happen, so I made `CollectionName` non-nullable, so the method should always return some name.

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreRecordCollection.cs (hunk 11 above)
Comment (dmytrostruk) [reply dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreRecordCollection.cs:130]: Added a possibility to specify API key in `options` and added XML documentation for `HttpClient` that it can also be included via `DefaultRequestHeaders` property. Also added tests to verify it is used correctly.

**Review by markwallace-microsoft — APPROVED** [review]
